File: AlgoExpert/2. Medium/C++/Spiral Traverse/Spiral Traverse [O(n)].cpp
#include <iostream>

using namespace std;

// O(n) time | O(n) space - where n is the total number of eleme

void spiralFill(vector<vector<int>> array, int startRow, int endRow, int startCol, int endCol, vector<int> &result) {

	if (startRow > endRow || startCol > endCol) {
		return;
	}

	for (int col = startCol; col <= endCol; col++) {
		result.push_back(array[startRow][col]);
	}

	for (int row = startRow + 1; row <= endRow; row++) {
		result.push_back(array[row][endCol]);
	}

	for (int col = endCol - 1; col >= startCol; col--) {
		if (startRow == endRow)
			break;
		result.push_back(array[endRow][col]);
	}

	for (int row = endRow - 1; row >= startRow + 1; row--) {
		if (startCol == endCol)
			break;
		result.push_back(array[row][startCol]);
	}

	spiralFill(array, startRow + 1, endRow - 1, startCol + 1, endC
}

File: AlgoExpert/2. Medium/Python/Spiral Traverse/Spiral Traverse [O(n)].py
# O(n) time | O(n) space - where n is the total number of elemen

def spiralTraverse(array):
	result = []
	startRow, endRow = 0, len(array) - 1
	startCol, endCol = 0, len(array[0]) - 1

	while startRow <= endRow and startCol <= endCol:
		for col in range(startCol, endCol + 1):
			result.append(array[startRow][col])

		for row in range(startRow + 1, endRow + 1):
			result.append(array[row][endCol])

		for col in reversed(range(startCol, endCol)):
			if startRow == endRow:
				break
			result.append(array[endRow][col])

		for row in reversed(range(startRow + 1, endRow)):
			if startCol == endCol:
				break
			result.append(array[row][startCol])

		startRow += 1
		endRow -= 1
		startCol += 1
		endCol -= 1
	return result